tests/core_tests/chaingen.cpp

```diff
@@ -55,6 +61,126 @@ using namespace epee;
+   * for the use only in the test_generator::fill_nonce() function,
+   * which requires blockchain object to correctly compute PoW on HF12+ blocks
+   * as the mining function requires it to obtain a valid seedhash.
+   */
+  class TestDB: public cryptonote::BaseTestDB
+  {
+  private:
+    struct block_t
+    {
+      cryptonote::block bl;
+      crypto::hash hash;
+    };
+
+  public:
+    TestDB() { m_open = true; }
+
+    virtual void add_block( const cryptonote::block& blk
+        , size_t block_weight
+        , uint64_t long_term_block_weight
+        , const cryptonote::difficulty_type& cumulative_difficulty
+        , const uint64_t& coins_generated
+        , uint64_t num_rct_outs
+        , const crypto::hash& blk_hash
+    ) override
+    {
+      blocks.push_back({blk, blk_hash});
+    }
+
+    virtual uint64_t height() const override { return blocks.empty() ? 0 : blocks.size() - 1; }
+
+    // Required for randomx
+    virtual crypto::hash get_block_hash_from_height(const uint64_t &height) const override
+    {
+      if (height < blocks.size())
+      {
+        MDEBUG("Get hash for block height: " << height << " hash: " << blocks[height].hash);
+        return blocks[height].hash;
+      }
+
+      MDEBUG("Get hash for block height: " << height << " zero-hash");
+      crypto::hash hash = crypto::null_hash;
+      *(uint64_t*)&hash = height;
+      return hash;
+    }
+
+    virtual crypto::hash top_block_hash(uint64_t *block_height = NULL) const override
+    {
+      const uint64_t h = height();
+      if (block_height != nullptr)
+      {
+        *block_height = h;
+      }
+
+      return get_block_hash_from_height(h);
+    }
+
+    virtual cryptonote::block get_top_block() const override
+    {
+      if (blocks.empty())
+      {
+        cryptonote::block b;
+        return b;
+      }
+
+      return blocks[blocks.size()-1].bl;
+    }
+
+    virtual void pop_block(cryptonote::block &blk, std::vector<cryptonote::transaction> &txs) override { if (!blocks.empty()) blocks.pop_back(); }
+    virtual void set_hard_fork_version(uint64_t height, uint8_t version) override { if (height >= hf.size()) hf.resize(height + 1); hf[height] = version; }
+    virtual uint8_t get_hard_fork_version(uint64_t height) const override { if (height >= hf.size()) return 255; return hf[height]; }
+
+  private:
+    std::vector<block_t> blocks;
+    std::vector<uint8_t> hf;
+  };
+
+}
+
+static std::unique_ptr<cryptonote::Blockchain> init_blockchain(const std::vector<test_event_entry> & events, cryptonote::network_type nettype)
+{
+  std::unique_ptr<cryptonote::Blockchain> bc;
+  v_hardforks_t hardforks;
+  cryptonote::test_options test_options_tmp{nullptr, 0};
+  const cryptonote::test_options * test_options = &test_options_tmp;
+  if (!extract_hard_forks(events, hardforks))
+  {
+    MDEBUG("Extracting hard-forks from blocks");
+    extract_hard_forks_from_blocks(events, hardforks);
+  }
```

Comment:
Is there a case where this is needed ? There are always test_options in core_tests (possibly empty, but in all cases the blocks have to match it).

Reply:
For the blockchain object construction I need to specify the hard forks so I extract the information from the `events`. In the `init_blockchain` I have no access to `test_options` specified in `core_tests` so I am using this approach.

In the Trezor tests I stored hard fork as a specific event (for some reason). As a fallback for tests not doing that I am reconstructing the hard forks from the blocks stored in the `events`.

Reply:
OK. Thanks for taking the time to find out and explain.